Employee directory part 2/?  - MVP + better styling

ahweigl opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
https://github.com/sandboxnu/mfa-form-automator/issues/323

## Description
- functional directory to edit and delete ('deactivate') employees WOOHOO! 
- better styling and not hard coded (but took a LOT of liberties with design)
- editing actually changes the employee, deleting happens by setting (new) `isActive` field on the employee to `false`, and not including it in the findAll response but not fr deleting it 

Whats NOT done: 
- changing scopes 
- adding employees 
- ux improvements and possibly better layout of page to include position/departments more seamlessly 
- pagination of employees
- filtering of employee directory

<!--- Describe your changes in detail -->

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->
we need emp dir
Closes [#323 ]

## How has this been tested?
UNTESTED :FEAR:
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):
see vid in slack 

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
^hopefully not

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

ahweigl commented on apps/web/src/components/ActiveFormList.tsx:
```diff
@@ -132,6 +132,7 @@ export const ActiveFormList = ({ title }: { title: string }) => {
                 searchQuery={searchQuery}
                 setSearchQuery={setSearchQuery}
                 setSortOption={setSortOption}
+                placeholder="Search for forms"
```
todo change to 'search forms'

ahweigl replied:
same w other usages

ahweigl commented on apps/web/src/context/EmployeesContext.tsx:
```diff
@@ -0,0 +1,43 @@
+  return (
+    <EmployeesContext.Provider
+      value={{
+        // TODO
```
todo fix type erorr here

ahweigl commented on apps/web/src/pages/employee-directory.tsx:
```diff
@@ -1,182 +1,178 @@
+  const [searchQuery, setSearchQuery] = useState('');
+  // TODO sorting 
+  const [sortOption, setSortOption] = useState<SortBy>(SortBy.NAME_DESC);
+  const { employees, isLoading, error } = useEmployeesContext();
```
todo add loading state, this stopped working

elvincheng3 commented on apps/web/src/components/ConfirmEmployeeChangesModal.tsx:
```diff
@@ -0,0 +1,147 @@
+  positions,
+}: ConfirmEmployeeChangesModalProps) => {
+  // @ts-ignore - position exists on employee but not in type
+  const currentDepartment = employee.position?.department;
```
there is a ts error because we're using the wrong type, we should be using `EmployeeSecureEntityHydrated` for entities with sensitive info.

elvincheng3 commented on apps/web/src/components/ConfirmEmployeeChangesModal.tsx:
```diff
@@ -0,0 +1,147 @@
+  const currentDepartment = employee.position?.department;
+  // @ts-ignore - position exists on employee but not in type
+  const currentPosition = employee.position;
+  const newDepartment = departments.find((d) => d.id === selectedDepartment);
```
if we have the selected department and position id already, why don't we pass in only the position and department we need instead of passing in every position and department as props?

elvincheng3 commented on apps/web/src/components/DeleteEmployeeModal.tsx:
```diff
@@ -0,0 +1,114 @@
+import { CloseIcon } from '@web/static/icons';
+
+// TODO docs
+export const DeleteEmployeeModal = ({
```
hoping kai can merge his PR in first; if he does there is an abstracted modal component for deleting things that can be used instead of making a new delete modal

elvincheng3 commented on apps/web/src/pages/employee-directory.tsx:
```diff
@@ -1,182 +1,322 @@
+  const [searchQuery, setSearchQuery] = useState('');
+  // TODO sorting
+  const [sortOption, setSortOption] = useState<SortBy>(SortBy.NAME_DESC);
+  const { employees, isLoading, error } = useEmployeesContext();
```
we should not create a new context that wraps a react query, we should use the react query in the component instead. it should expose all the same props you're using here

elvincheng3 commented on apps/web/src/pages/_app.tsx:
```diff
@@ -217,10 +218,21 @@ export default function App({
-    appProps.router.pathname.includes(instancesDirectoryPath) ||
-    appProps.router.pathname.includes(employeeDirectoryPath)
-  ) {
+  if (appProps.router.pathname.includes(employeeDirectoryPath)) {
```
don't need this since we don't need a separate context for employees

elvincheng3 commented on apps/web/src/pages/employee-directory.tsx:
```diff
@@ -1,182 +1,322 @@
+    fetchPositions();
+  }, [selectedDepartment]);
 
+  const handleEditClick = (employee: EmployeeBaseEntity) => {
```
using `EmployeeSecureEntityHydrated` will get rid of the ts-errors. you'll need to pass a `secure=true` param when fetching employees to get the secure entities and not the base employee entities

elvincheng3 commented on apps/web/src/pages/employee-directory.tsx:
```diff
@@ -1,182 +1,322 @@
+                </Box>
+                <Box flex={2}>
+                  {editingEmployee === employee.id ? (
+                    <select
```
let's use the chakra-react-select component instead

elvincheng3 replied:
or refer to other usages of <select> in the codebase for styling

elvincheng3 commented on apps/web/src/pages/employee-directory.tsx:
```diff
@@ -1,182 +1,322 @@
+                      ))}
+                    </select>
+                  ) : (
+                    // @ts-ignore - position exists on employee but not in type
```
should be resolved when other comments are addressed

elvincheng3 commented on apps/server/prisma/schema.prisma:
```diff
@@ -68,10 +68,11 @@ model Employee {
   firstName     String        @db.VarChar(255)
   lastName      String        @db.VarChar(255)
-  email         String        @unique @db.VarChar(255)
+  email         String        @db.VarChar(255)
```
shouldn't emails be unique?

elvincheng3 commented on apps/server/prisma/seeder.ts:
```diff
@@ -175,7 +175,8 @@ export class Seeder {
         scope: scope,
-      },
+        isActive: true,
+      } as any,
```
let's fix this typing

elvincheng3 commented on apps/server/src/employees/dto/update-employee.dto.ts:
```diff
@@ -3,7 +3,7 @@ import { CreateEmployeeDto } from './create-employee.dto';
 
 export class UpdateEmployeeDto extends PartialType(
-  OmitType(CreateEmployeeDto, ['email', 'password'] as const),
+  OmitType(CreateEmployeeDto, ['email'] as const),
```
why are we including password now? technically everyone should be using azure, so the passwords wouldn't even be used

elvincheng3 commented on apps/server/src/employees/employees.service.ts:
```diff
@@ -40,16 +40,26 @@ export class EmployeesService {
    * @returns the created employee, hydrated
    */
   async create(createEmployeeDto: CreateEmployeeDto) {
+    // manually make sure email is unique (since we are not using unique constraint in the database due to disabled flag)
```
i think it makes more sense to re-enable an account with the same email instead of making a new acct

elvincheng3 commented on apps/server/src/employees/employees.service.ts:
```diff
@@ -74,10 +84,14 @@ export class EmployeesService {
+          orderBy: orderBy(sortBy, true),
+          where: {
+            isActive: true,
+          } as any,
```
why do we need type casting?

elvincheng3 commented on apps/server/src/employees/employees.service.ts:
```diff
@@ -86,6 +100,10 @@ export class EmployeesService {
+          orderBy: orderBy(sortBy, true),
+          where: {
+            isActive: true,
+          } as any,
```
type casting

elvincheng3 commented on apps/server/src/employees/employees.service.ts:
```diff
@@ -97,9 +115,19 @@ export class EmployeesService {
+      orderBy: orderBy(sortBy, true),
+      where: {
+        isActive: true,
+      } as any,
```
type casting

elvincheng3 commented on apps/server/src/employees/employees.service.ts:
```diff
@@ -248,13 +280,28 @@ export class EmployeesService {
+      },
+      data: {
+        isActive: false,
+      } as any,
```
type casting

elvincheng3 commented on apps/web/openapi-ts-error-1746068469655.log:
```diff
@@ -0,0 +1,11 @@
+Error requesting http://localhost:8080/api-json 
+fetch failed
+ResolverError: Error requesting http://localhost:8080/api-json 
```
let's delete this file

elvincheng3 commented on apps/web/src/components/ConfirmEmployeeChangesModal.tsx:
```diff
@@ -0,0 +1,281 @@
+
+  const currentDepartment = employee.position?.department;
+  const currentPosition = employee.position;
+  const currentScope = (employee.scope as Scope) || Scope.BASE_USER;
```
do we need the cast to Scope here?

kaiyangzheng replied:
yes

elvincheng3 commented on apps/web/src/components/editDepartment/ModifyDepartmentCard.tsx:
```diff
@@ -71,6 +108,12 @@ export const ModifyDepartmentCard = ({
         queryKey: positionsControllerFindAllQueryKey(),
       });
       refreshUser();
+      toaster.create({
```
if we set loading to false in error state, should we do that in success too?

elvincheng3 commented on apps/web/src/components/editDepartment/ModifyDepartmentCard.tsx:
```diff
@@ -99,30 +148,36 @@ export const ModifyDepartmentCard = ({
+    try {
+      await Promise.all(
+        department.positions.map((position) => {
+          return updatePosition.mutateAsync({
```
works for now, but it might be worth having this logic happen on the backend instead and hiding it from the client

kaiyangzheng replied:
:died:

ahweigl replied:
ELVIN NO

elvincheng3 commented on apps/web/src/pages/employee-directory.tsx:
```diff
@@ -1,111 +1,447 @@
+import { PlusIcon, UserProfileAvatar } from '@web/static/icons';
+import { useState, useEffect } from 'react';
+import { FiEdit2 } from 'react-icons/fi';
+import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
```
unused import

elvincheng3 commented:
bug: only one employee should be able to be assigned to a position?

elvincheng3 commented:
bug: when creating employee, should validate against azure that email is valid

ahweigl commented:
cursor generated PR review: 
Summary of Changes
1. Database & Entity Changes
Added isActive boolean field to the Employee model in the database schema with a default value of true
Created a migration script for this schema change
Updated entity classes (EmployeeBaseEntity and EmployeeSecureEntity) to include the isActive field with proper Swagger documentation
2. Backend API Enhancements
Employee Service
Implemented the soft deletion pattern with a new deactivate method that sets an employee's isActive to false
Modified findAll and findAllSecure methods to accept an options object with limit, sortBy, and isActive parameters
Added filtering by isActive: true to all employee query methods including findOne and authentication methods
Added more robust email uniqueness validation during employee creation
Added sorting capabilities using a utility function
Employee Controller
Updated remove endpoint to call deactivate instead of the hard-delete method
Added a new disabled endpoint to list deactivated employees
Enhanced the findAll endpoint to support sorting
Added better error handling for duplicate emails
Updated the input parameter structure for service methods
3. Frontend UI Improvements
Employee Directory Page
Complete overhaul of the employee directory UI:
Replaced mock data with actual API calls
Implemented real-time filtering and sorting capabilities
Added ability to view both active and deactivated employees
Added employee editing functionality with confirmation modals
Added department and position management modals
Added interface for creating new employees
Implemented proper error handling with toast notifications
New Components
ConfirmEmployeeChangesModal: Shows changes before saving
DeleteConfirmModal: Confirms before deactivating employees
SearchAndSort: Unified component for filtering and sorting lists
Enhanced department and position management modals
4. Authentication & Security
Updated authentication flows to check for isActive: true when validating users
Added validation to prevent users from deactivating their own accounts
Implemented proper role-based access control for admin operations
5. API Client & Types
Updated generated API clients to match new endpoints and parameters
Enhanced TypeScript types to include the isActive field and sorting options
Code Quality Improvements
Cleaner State Management: Better organization of component state
Enhanced Error Handling: More robust error handling with user feedback
Better Type Safety: Improved TypeScript typing throughout the application
Better API Structure: More flexible API endpoints with proper parameter handling
Improved Sorting and Filtering: Server-side sorting and filtering capabilities
PR Review Comments
Positive Aspects:
Excellent implementation of the soft deletion pattern
Comprehensive approach touching database, API, and UI layers
Well-designed UI with proper feedback mechanisms
Good type safety and error handling
Addition of a way to view/manage deactivated employees
Suggestions for Improvement:
Consider adding a reactivation feature for deactivated employees
The employeesService.findOne method could have an option to find deactivated employees
Add tests specifically for the soft deletion and reactivation flows
Consider adding a way to permanently delete employees for GDPR/data retention compliance
Some methods in the employee service could be refactored to reduce code duplication
Potential Issues:
Need to ensure that all authentication checks properly verify isActive: true
Sessions for deactivated users should be invalidated
Consider adding logging when employees are deactivated for audit purposes
The UI for managing deactivated employees could use some visual distinction from active employees
Conclusion
This branch introduces a well-implemented soft deletion pattern for the employee management system. The changes are comprehensive, touching all layers of the application from the database schema to the frontend UI. The implementation follows best practices for soft deletion and provides a good user experience.
The most significant improvements include:
The ability to deactivate employees instead of permanently deleting them
Enhanced UI for managing employees with proper feedback mechanisms
Improved API with more flexible parameters and better error handling
Complete replacement of mock data with real API calls
Overall, this is a solid implementation that significantly improves the employee management functionality of the application.

elvincheng3 commented on apps/web/src/pages/employee-directory.tsx:
```diff
@@ -392,6 +447,18 @@ function EmployeeDirectory() {
+      path: { id: employeeId },
+      body: {
+        isActive: true,
+      },
```
nit: error handling/toast

kaiyangzheng replied:
erro handling is defined in the mutation

elvincheng3 commented on apps/web/src/pages/form-instance/[id]/edit/EditGuard.tsx:
```diff
@@ -0,0 +1,35 @@
+  }
+
+  // Render children if user has permission
+  return <>{children}</>;
```
nit: do we need the surrounding tags?
